#202 Step 1 - Add `Roles and permissions` section UI on GroupSettings page

SebinSong opened the pull request:
Step 1 of #202 Todo list: [here](https://github.com/okTurtles/group-income/issues/202#issuecomment-2567190831)

NOTE: This PR is UI implementation only. The data presented in the table are fake.

[ Desktop ]

<img src='https://github.com/user-attachments/assets/9fdd112c-ffc9-4ccc-b8e6-c45443d6f1a3' width='560'>

[ Mobile ]

<img src='https://github.com/user-attachments/assets/d7c57f91-a05e-4ce2-a7cf-50a50de37e28' width='340'>

SebinSong committed:
build boiler-plate RolesAndPermissions.vue

SebinSong committed:
add some GROUP_PERMISSIONS related constants and define some presets

SebinSong committed:
build boilderplate PermissionTableRow.vue

SebinSong committed:
prepare dummy permissions data / displayname logic for roles and permissions

SebinSong committed:
build PermissionActionMenu.vue for table menu

SebinSong committed:
work on ViewPermissions.vue

SebinSong committed:
Merge remote-tracking branch 'origin/master' into sebin/task/#202-roles-and-permissions-step1

SebinSong committed:
build ViewPermissions tooltip

SebinSong committed:
work on feedbacks / enhance ViewPermissions.vue overlay UI

SebinSong committed:
make sure the table looks presentable on mobile width

SebinSong committed:
design bug-fixes for ViewPermossions.vue overlay UI

SebinSong committed:
fix css linter err

SebinSong committed:
add some comments

SebinSong committed:
fix the cypress err

SebinSong committed:
fix the group-setting spec failure

taoeffect commented on frontend/views/pages/GroupSettings.vue:
```diff
@@ -80,6 +80,8 @@ page.c-page
           data-test='saveBtn'
         ) {{ L('Save changes') }}
 
+  roles-and-permissions
+
```
Can we have this section displayed only in development mode since it doesn't work and hide it in production?

SebinSong replied:
Sure, updated in [here](https://github.com/okTurtles/group-income/pull/2507/commits/4eaec1e0432e4c0a3dec6b90f3d2b8cb7dada70a)

taoeffect reviewed (changes requested):
Nice work @SebinSong! Preliminary review ready (I haven't had a chance to do an in-depth review today)

SebinSong committed:
show UI only in dev mode

SebinSong committed:
Merge remote-tracking branch 'origin/master' into sebin/task/#202-roles-and-permissions-step1

SebinSong committed:
better table structure in mobile-size screen

SebinSong commented:
Just made some additional updates:

- Better table structure for mobile size screen. (Combine 'user' and 'role' columns for smaller space)
- No need to expose 'See all' button when there is only one item to show.

<img src='https://github.com/user-attachments/assets/6e9f2710-aa44-45c8-8972-d80c51734619' width='342'>

SebinSong committed:
Merge remote-tracking branch 'origin/master' into sebin/task/#202-roles-and-permissions-step1

SebinSong committed:
update comment

SebinSong committed:
Merge remote-tracking branch 'origin/master' into sebin/task/#202-roles-and-permissions-step1

SebinSong committed:
Merge remote-tracking branch 'origin/master' into sebin/task/#202-roles-and-permissions-step1

SebinSong committed:
Merge remote-tracking branch 'origin/master' into sebin/task/#202-roles-and-permissions-step1

taoeffect commented on frontend/model/contracts/shared/constants.js:
```diff
@@ -59,6 +59,45 @@ export const STREAK_MISSED_PAYMENTS = 1
+export const GROUP_PERMISSIONS = {
+  VIEW_PERMISSIONS: 'view-permissions',
+  REMOVE_PERMISSIONS: 'remove-permissions',
+  DELEGATE_PERMISSIONS: 'delegate-permissions',
```
What is the idea as far as the difference between `REMOVE_PERMISSIONS` and `DELEGATE_PERMISSIONS`?

I think if you have the ability to give permissions (`DELEGATE_PERMISSIONS`), it should imply the ability to take them away (`REMOVE_PERMISSIONS`), what do you think?

taoeffect commented on frontend/model/contracts/shared/constants.js:
```diff
@@ -59,6 +59,45 @@ export const STREAK_MISSED_PAYMENTS = 1
+  ],
+  MODERATOR: [
+    GP.VIEW_PERMISSIONS
+  ]
```
`MODERATOR` AND `MODERATOR_DELEGATOR` should have the same exact permissions, except that `MODERATOR_DELEGATOR` should have `GP.DELEGATE_PERMISSIONS`.

taoeffect commented on frontend/views/containers/group-settings/roles-and-permissions/PermissionActionMenu.vue:
```diff
@@ -0,0 +1,84 @@
+        }
+      ]
+
+      return list.filter(option => option.enabledFor.includes(this.role))
```
Instead of checking the roles here, we should be checking the permissions.

To be able to edit or remove permissions, you need both `DELEGATE_PERMISSIONS`, and then you can only edit permissions that you have.

By combining `DELEGATE_PERMISSIONS` and `REMOVE_PERMISSIONS`, we only need to check for `DELEGATE_PERMISSIONS`.

SebinSong replied:
Okay, will remove`REMOVE_PERMISSIONS` and use `DELEGATE_PERMISSIONS` everywhere.

taoeffect commented on frontend/views/containers/group-settings/roles-and-permissions/RolesAndPermissions.vue:
```diff
@@ -0,0 +1,163 @@
+    role: GROUP_ROLES.MODERATOR,
+    permissions: GROUP_PERMISSIONS_PRESET.MODERATOR
+  }
+]
```
Can we see one for Custom too just to see what it looks like?

taoeffect reviewed (changes requested):
Great work @SebinSong!

Left a few comments and tested it, seems to work well on both desktop and mobile!

Can you also merge `master` since it's been a while?

SebinSong committed:
Merge remote-tracking branch 'origin/master' into sebin/task/#202-roles-and-permissions-step1

SebinSong committed:
Merge remote-tracking branch 'origin/master' into sebin/task/#202-roles-and-permissions-step1

SebinSong committed:
update per change requests

SebinSong commented:
@taoeffect 

Updated the PR per your feedbacks.

### 1.
I removed `REMOVE_PERMISSIONS` as advised. (I think it's more clear to me now what `DELEGATE_PERMISSIONS` means). Also, as per below comment you left in the issue:

> However, it would not mean that the delegator could delegate the delegate-permission to others. Only the admin would be able to do this. In other words, if the admin granted me moderator (delegator) role, I would only be able to grant moderator permissions to others, but not moderator (delegator).

I created a `ASSIGN_DELEGATOR` permission entry and added it to `ADMIN` preset only.

### 2.
Added a dummy user for a custom role here:

<img src='https://github.com/user-attachments/assets/75b652f1-b96a-44f0-bc46-2f4b1a3e2446' width='520'>

taoeffect reviewed (approved):
Great work @SebinSong!
